File: TidbitsTrivia/Core/Data/DifficultyOverlay.swift
import Foundation

/// F3 — the derived difficulty overlay (Wikipedia pageviews → 1..5 per subject),
/// loaded from the bundled `difficulty.json`. Additive: the corpus is untouched;
/// the Ladder mode sorts by this and weights its scoring by it. A subject not in
/// the map (or a live-generated question) defaults to 3 (middle).
nonisolated final class DifficultyOverlay: @unchecked Sendable {
    static let shared = DifficultyOverlay()

    private let map: [String: Int]

    init() {
        guard let url = Bundle.main.url(forResource: "difficulty", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let dict = root["difficulty"] as? [String: Int] else {
            map = [:]
            return
        }
        map = dict
    }

    /// Difficulty 1 (best-known) … 5 (obscure). `title` is the display title; the
    /// overlay is keyed by the underscored Wikipedia title.
    func difficulty(forTitle title: String) -> Int {
        map[title.replacingOccurrences(of: " ", with: "_")] ?? 3
    }

    func difficulty(for question: Question) -> Int { difficulty(forTitle: question.sourceTitle) }
}

File: windows/Tidbits.Core/Data/DifficultyOverlay.cs
using System.Text.Json;
using Tidbits.Core.Models;

namespace Tidbits.Core.Data;

/// F3 derived-difficulty overlay (Wikipedia pageviews → 1..5 per subject), from
/// the shared difficulty.json. Additive — the corpus is untouched; Ladder mode
/// sorts + weights scoring by it. A subject not in the map (or a live-generated
/// question) defaults to 3. Port of Core/Data/DifficultyOverlay.swift.
public sealed class DifficultyOverlay
{
    private readonly Dictionary<string, int> _map;

    public DifficultyOverlay(Dictionary<string, int>? map = null) => _map = map ?? new();

    public static DifficultyOverlay Load(Stream json)
    {
        using var doc = JsonDocument.Parse(json);
        var map = new Dictionary<string, int>();
        if (doc.RootElement.TryGetProperty("difficulty", out var d) && d.ValueKind == JsonValueKind.Object)
            foreach (var p in d.EnumerateObject())
                if (p.Value.TryGetInt32(out var v)) map[p.Name] = v;
        return new DifficultyOverlay(map);
    }

    /// 1 (best-known) … 5 (obscure). Keyed by the underscored Wikipedia title.
    public int DifficultyForTitle(string title) => _map.GetValueOrDefault(title.Replace(' ', '_'), 3);

    public int DifficultyFor(Question q) => DifficultyForTitle(q.SourceTitle);
}
